src/server/acl/user.h

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <cstdint>
+#include <limits>
+#include <optional>
+#include <string>
+#include <string_view>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/hash/hash.h"
+
+namespace dfly {
+
+class CommandId;
+
+bool CheckIfCommandAllowed(uint64_t command_id, const CommandId& command);
+bool CheckIfAclCategoryAllowed(uint64_t command_id, const CommandId& command);
+
+namespace AclCat {
+/* There are 21 ACL categories as of redis 7
+ *
+ * bit 0: admin
+ * bit 1: bitmap
+ * bit 2: blocking
+ * bit 3: connection
+ * bit 4: dangerous
+ * bit 5: geo
+ * bit 6: hash
+ * bit 7: list
+ * bit 8: set
+ * bit 9: string
+ * bit 10: sorttedset
+ * bit 11: hyperloglog
+ * bit 12: streams
+ * bit 13: fast
+ * bit 14: slow
+ * bit 15: key-space
+ * bit 16: pubsub
+ * bit 17: read
+ * bit 18: write
+ * bit 19: scripting
+ * bit 20: transaction
+ *
+ * The rest of the bitfield, will contain special values like @all
+ *
+ * bits 21..31: tba
+ */
+inline constexpr uint32_t ACL_CATEGORY_KEYSPACE = 1ULL << 0;
+inline constexpr uint32_t ACL_CATEGORY_READ = 1ULL << 1;
+inline constexpr uint32_t ACL_CATEGORY_WRITE = 1ULL << 2;
```

Comment:
So you have a dedicated namespace with global constants, reminds me of an.... enum (class) ! 😄

Reply:
yes but enum class does not support implicit conversions between objects and the underlying type which it would make it a nightmare to use without casting on each operation. The other idea was enum + namespace which now that I am thinking it it's a good fit!

Reply:
yes but this requires explicit conversions it's used. class enums protect against it.

for example see https://godbolt.org/z/xvK5MMb4v